proto/src/simulator/component.proto

```diff
@@ -0,0 +1,139 @@
+// or if the client explicitly requests a component summary. (E.g. for WAILA.)
+//
+// It's *not* saved to disk, as there is quite a bit of data which the client never needs. Those protobuf
+// variants exist only inside the core module.
+syntax = "proto3";
+package org.eln2.proto.simulator;
+
+// Each field is only filled in when it makes sense, i.e. a crash report will include everything,
+// as might a watchdog firing or WAILA-induced state request, but an AddComponent request won't
+// include the (unknown) electrical state. It might include thermal state.
+message Component {
+	// This value is monotonically increasing for a given server, and never reused.
+	int64 id = 1;
+
+	// This is the computed electrical and thermal state, usually sent server->client.
+	ElectricalState electrical_state = 2;
+	ThermalState thermal_state = 3;
+
+	// These are initialization parameters, sent client->server.
+	Watchdog watchdog = 4;
+	ThermalParams thermal = 5;
+	Location location = 6;
+	// In general, default to adding components to this list (handling them core-side) rather than
+	// emulating them client-side. E.g, a lightbulb is better simulated in core than on the client.
+	// This is both for performance (it's much easier to multithread Core than a minecraft server),
+	// and for code de-duplication.
+	oneof type {
+		// Circuit primitives:
+		VoltageSource voltage_source = 10;
+		CurrentSource current_source = 11;
+		Resistor resistor = 12;
+		Inductor inductor = 13;
+		Capacitor capacitor = 14;
+		// Cables:
+		Resistor simple_cable = 20;
+		// Lights:
+		IncandescentLight incandescent_light = 21;
+		FluorescentLight fluorescent_light = 22;
+	}
+}
+
+// Standard thermal parameters.
+message ThermalParams {
+	float thermal_mass = 1;
+	// TODO: Fill in.
+}
+
+// Standard thermal state.
+message ThermalState {
+	float temperature = 1;
+}
+
+// Standard electrical state.
+message ElectricalState {
+	message Pin {
+		float voltage = 1;
+		float current = 2;
+	}
+	// Pin 0, pin 1, pin 2, etc.
+	repeated Pin pin = 1;
+}
+
+// Standard watchdogs.
+//
+// The simulator won't take direct action in response to a component exceeding its physical limits. Instead it will
+// notify the client that this has happened. The client can then make its own decisions based on gameplay integration.
+//
+// Not all watchdog limits make sense for all components.
+// For components with multiple pins, this applies to all of them. Any watchdogs that shouldn't are defined per-component.
+message Watchdog {
+	// 'Low' values should typically be negative, if set.
+	float voltage_low = 1;
+	float voltage_high = 2;
+	float current_low = 3;
+	float current_high = 4;
+	float temperature_low = 5;  // Uh... for a chiller, maybe? In kelvin, of course.
+	float temperature_high = 6;
+}
+
+// In-world positioning.
+// This has no bearing on the simulation, unless perhaps if we implement a deeper thermal simulation.
+// Currently it's purely informational.
+message Location {
```

Comment:
This won't be able to handle SixNode. It doesn't store orientation data. My current guess is that if we were to use this, it would be inductor sharing, which is within a block. Also, storing as double may be advisable.

Reply:
That's a fair point, but OTOH it's meant to be advisory; it has no function inside the server.

I don't know how sixnode will end up looking here, given the SCM system I'm aiming for. Will add once that is clear, so probably much later.